toolchain/check/handle_class.cpp

```diff
@@ -172,15 +183,79 @@ auto HandleClassDefinitionStart(Context& context,
   return true;
 }
 
-auto HandleAdaptIntroducer(Context& /*context*/,
+auto HandleAdaptIntroducer(Context& context,
                            Parse::AdaptIntroducerId /*node_id*/) -> bool {
+  context.decl_state_stack().Push(DeclState::Adapt);
   return true;
 }
 
-auto HandleAdaptDecl(Context& context, Parse::AdaptDeclId /*node_id*/) -> bool {
-  auto adapted_expr_id = context.node_stack().PopExpr();
-  (void)adapted_expr_id;
-  // TODO: Process the `adapt` declaration.
+auto HandleAdaptDecl(Context& context, Parse::AdaptDeclId node_id) -> bool {
+  auto [adapted_type_node, adapted_type_expr_id] =
+      context.node_stack().PopExprWithNodeId();
+
+  // Process modifiers. `extend` is permitted, no others are allowed.
+  LimitModifiersOnDecl(context, KeywordModifierSet::Extend,
+                       Lex::TokenKind::Adapt);
+  auto modifiers = context.decl_state_stack().innermost().modifier_set;
+  context.decl_state_stack().Pop(DeclState::Adapt);
+
+  auto enclosing_class_decl = context.GetCurrentScopeAs<SemIR::ClassDecl>();
+  if (!enclosing_class_decl) {
+    CARBON_DIAGNOSTIC(AdaptOutsideClass, Error,
+                      "`adapt` declaration can only be used in a class.");
+    context.emitter().Emit(node_id, AdaptOutsideClass);
+    return true;
+  }
+
+  auto& class_info = context.classes().Get(enclosing_class_decl->class_id);
+  if (class_info.adapt_id.is_valid()) {
+    CARBON_DIAGNOSTIC(AdaptRepeated, Error,
+                      "Multiple `adapt` declarations in class.");
+    CARBON_DIAGNOSTIC(AdaptPrevious, Note,
+                      "Previous `adapt` declaration is here.");
```

Comment:
Done in a couple of places. This one is a bit awkward because we have extra text in the base case, but I think it's still worth factoring out.

Reply:
Yeah, we do similar elsewhere, although maybe we should provide a better way to achieve it. You might consider llvm::StringLiteral instead of std::string though, given it's constant.